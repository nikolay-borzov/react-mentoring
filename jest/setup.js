import promiseFinally from 'promise.prototype.finally'

import Enzyme, { shallow, render, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// Bring '.finally' to Promise (only for Node 8). Will be available in Node 10
promiseFinally.shim()

Enzyme.configure({ adapter: new Adapter() })

// Make Enzyme functions available in all test files without importing
global.shallow = shallow
global.render = render
global.mount = mount
