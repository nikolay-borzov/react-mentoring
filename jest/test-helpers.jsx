import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

/**
 * Creates test case for snapshot testing
 * @param {Function} getComponent Function returning component to test snapshot
 * @param {string} [testName='renders correctly']
 */
export const itRendersCorrectly = (
  getComponent,
  testName = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent())

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test case to check whether component contains another component
 * @param {Function} getComponent Function returning parent component
 * @param {string} componentDisplayName Child component display name
 * @param {number} [expectedCount=1] Expected number of children components
 * @param {Object} withProps First child component properties map
 */
export const itContainsComponent = (
  getComponent,
  componentDisplayName,
  expectedCount = 1,
  withProps
) => {
  it(`contains ${componentDisplayName} component`, () => {
    const wrapper = shallow(getComponent())

    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)

    if (withProps) {
      expect(children.first().props()).toEqual(withProps)
    }
  })
}
