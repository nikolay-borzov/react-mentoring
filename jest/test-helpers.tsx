import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { runSaga as runSagaOriginal } from 'redux-saga'
import configureMockStore from 'redux-mock-store'
import { AnyAction } from 'redux';

/**
 * Creates test closure for snapshot testing
 * @deprecated Use itRendersCorrectlyShallow
 */
export const itRendersCorrectly = (
  getComponent: Function,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const tree = renderer.create(getComponent()).toJSON()

    expect(tree).toMatchSnapshot()
  })
}

/**
 * Creates test closure for snapshot testing (using Enzyme shallow)
 * @param getComponent Function returning component to test snapshot
 */
export const itRendersCorrectlyShallow = (
  getComponent: () => React.ReactElement,
  testName: string = 'renders correctly'
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())

    expect(wrapper).toMatchSnapshot()
  })
}

interface ItContainsComponentParams {
  testName?: string
  expectedCount?: number
  expectedProps?: {}
}

/**
 * Creates test closure to check whether component contains another component
 * @param getComponent Function returning parent component
 * @param componentDisplayName Child component display name
 * @param params Optional params
 * @param [params.testName=`contains ${componentDisplayName} component`]
 * @param [params.expectedCount=1] Expected number of children components
 * @param params.expectedProps First child component properties map
 */
export const itContainsComponent = (
  getComponent: () => React.ReactElement,
  componentDisplayName: string,
  { testName = `contains ${componentDisplayName} component`, expectedCount = 1, expectedProps }: ItContainsComponentParams = {}
) => {
  it(testName, () => {
    const wrapper = shallow(getComponent())
    const children = wrapper.find(componentDisplayName)

    expect(children.length).toBe(expectedCount)
    if (expectedProps) {
      expect(children.first().props()).toEqual(expectedProps)
    }
  })
}

/**
 * Creates mocked store
 */
export const getMockStore = () => configureMockStore([])

/**
 * Sets relative URL
 * @param {string} url
 */
export const setUrl = (url: string) => {
  window.history.pushState({}, '', url)
}

/**
 * Runs saga and collects dispatched actions
 */
export const runSaga = (saga: any, state = {}, arg1: any = undefined) => {
  const dispatched: AnyAction[] = []
  return runSagaOriginal(
    {
      dispatch: (action: AnyAction) => dispatched.push(action),
      getState: () => state
    },
    saga,
    arg1
  )
    .toPromise()
    .then(() => dispatched)
}
