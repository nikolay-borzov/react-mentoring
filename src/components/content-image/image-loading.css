@import '../../styles/variables.css';


/* https://github.com/tobiasahlin/SpinKit */
.image-spinner {
  width: 5rem;
  height: 5rem;
}

.image-spinner__bounce1,
.image-spinner__bounce2 {
  width: 100%;
  height: 100%;
  border-radius: 50%;
  background-color: var(--color-default);
  opacity: 0.6;
  position: absolute;
  top: 0;
  left: 0;

  animation: sk-bounce 2s infinite ease-in-out;
}

.image-spinner__bounce2 {
  animation-delay: -1s;
}

@keyframes sk-bounce {
  0%,
  100% {
    transform: scale(0);
  }
  50% {
    transform: scale(1);
  }
}
