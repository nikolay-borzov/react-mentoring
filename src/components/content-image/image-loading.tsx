import React from 'react'
import styled, { keyframes } from 'styled-components'

/* Animation from https://github.com/tobiasahlin/SpinKit */
const ImageSpinner = styled.div`
  width: 5rem;
  height: 5rem;
`

const bounce = keyframes`
 0%,
  100% {
    transform: scale(0);
  }
  50% {
    transform: scale(1);
  }
`

const ImageSpinnerBounce1 = styled.div`
  width: 100%;
  height: 100%;
  border-radius: 50%;
  background-color: var(--color-default);
  opacity: 0.6;
  position: absolute;
  top: 0;
  left: 0;

  animation: ${bounce} 2s infinite ease-in-out;
`
const ImageSpinnerBounce2 = styled(ImageSpinnerBounce1)`
  animation-delay: -1s;
`

export const ImageLoading: React.FunctionComponent = () => (
  <ImageSpinner>
    <ImageSpinnerBounce1 />
    <ImageSpinnerBounce2 />
  </ImageSpinner>
)
