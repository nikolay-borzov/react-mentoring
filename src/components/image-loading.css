.image-loading-indicator {
  position: relative;
  width: 100%;
  /* Keep aspect ratio */
  padding-top: 150%;
  background-color: rgba(0, 0, 0, 0.1);
}

/* https://github.com/tobiasahlin/SpinKit */
.image-spinner {
  width: 5em;
  height: 5em;

  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.image-spinner__bounce1,
.image-spinner__bounce2 {
  width: 100%;
  height: 100%;
  border-radius: 50%;
  background-color: var(--color-default);
  opacity: 0.6;
  position: absolute;
  top: 0;
  left: 0;

  animation: sk-bounce 2s infinite ease-in-out;
}

.image-spinner__bounce2 {
  animation-delay: -1s;
}

@keyframes sk-bounce {
  0%,
  100% {
    transform: scale(0);
  }
  50% {
    transform: scale(1);
  }
}
