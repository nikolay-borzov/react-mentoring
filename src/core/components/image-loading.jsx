import React from 'react'

export function ImageLoading() {
  return (
    <div className="image-loading-indicator">
      <div className="image-spinner">
        <div className="image-spinner__bounce1" />
        <div className="image-spinner__bounce2" />
      </div>
    </div>
  )
}
