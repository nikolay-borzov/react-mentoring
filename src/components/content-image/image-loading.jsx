import React from 'react'
import PropTypes from 'prop-types'

import './image-loading.css'

ImageLoading.propTypes = {
  className: PropTypes.string
}

ImageLoading.defaultProps = {
  className: ''
}

export function ImageLoading(props) {
  return (
    <div className={`image-loading-indicator ${props.className}`}>
      <div className="image-spinner">
        <div className="image-spinner__bounce1" />
        <div className="image-spinner__bounce2" />
      </div>
    </div>
  )
}
