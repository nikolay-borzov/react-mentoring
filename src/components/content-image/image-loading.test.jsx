import React from 'react'

import { itRendersCorrectly } from '../../../jest/test-helpers'

import { ImageLoading } from './image-loading'

describe('ImageLoading component', () => {
  itRendersCorrectly(() => <ImageLoading />)
})
