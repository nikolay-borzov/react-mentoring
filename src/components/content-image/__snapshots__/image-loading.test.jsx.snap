// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLoading component renders correctly 1`] = `
<div
  className="image-loading-indicator"
>
  <div
    className="image-spinner"
  >
    <div
      className="image-spinner__bounce1"
    />
    <div
      className="image-spinner__bounce2"
    />
  </div>
</div>
`;
